private/bufpkg/bufplugin/bufplugindocker/context.go

```diff
@@ -0,0 +1,91 @@
+// Copyright 2020-2022 Buf Technologies, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package bufplugindocker
+
+import (
+	"archive/tar"
+	"bytes"
+	"io"
+	"io/fs"
+	"time"
+)
+
+// createDockerContext builds a light-weight Docker context (tar file) from a given Dockerfile.
+// This is used to create images that don't rely on any other filesystem state to copy into the context.
+func createDockerContext(dockerfile io.Reader) (io.Reader, error) {
+	dockerfileBytes, err := io.ReadAll(dockerfile)
+	if err != nil {
+		return nil, err
+	}
+	var buffer bytes.Buffer
+	writer := tar.NewWriter(&buffer)
+	dockerfileInfo := &fileInfo{
+		name: "Dockerfile",
+		size: int64(len(dockerfileBytes)),
+		mode: 0644,
+	}
+	header, err := tar.FileInfoHeader(dockerfileInfo, "")
+	if err != nil {
+		_ = writer.Close()
+		return nil, err
+	}
+	if err := writer.WriteHeader(header); err != nil {
+		_ = writer.Close()
```

Comment:
Let's defer the `writer.Close` call as soon as the `writer` is constructed instead of manually calling it in all these places.

We'll also want to capture the error and we can do so with a named return value (you'll see this elsewhere too).

```go
func createDockerContext(dockerfile io.Reader) (_ io.Reader, retErr error) {
  ...
  writer := tar.NewWriter(&buffer)
  defer func() {
    retErr = multierr.Append(retErr, writer.Close()
  }()
  ...
}
```